api/src/main/java/org/apache/iceberg/expressions/ManifestEvaluator.java

```diff
@@ -340,6 +340,56 @@ public Boolean or(Boolean leftResult, Boolean rightResult) {
       return ROWS_MIGHT_MATCH;
     }
 
+    @Override
+    public <T> Boolean notStartsWith(BoundReference<T> ref, Literal<T> lit) {
+      int pos = Accessors.toPosition(ref.accessor());
+      PartitionFieldSummary fieldStats = stats.get(pos);
+
+      // Iceberg does not implement SQL 3-boolean logic. Therefore, for all null values, we have decided to
+      // return ROWS_MIGHT_MATCH in order to allow the query engine to further evaluate this partition, as
+      // null does not start with any non-null value.
+      if (fieldStats.containsNull() || fieldStats.lowerBound() == null) {
```

Comment:
`fieldStats.lowerBound() == null` is checked in the `if` below, so no need to duplicate that here.

Reply:
I don't think that we need the null check here. Null values do not match, but we can't tell whether _all_ the values are null or not from these stats. So all we can do is ignore this and check the bounds.

You should be able to just remove this `if` block entirely.

Reply:
Nevermind, this is correct as I mentioned above in the inclusive metrics evaluator. I'd probably change the comment and remove the part about 3-value logic.